### PR TITLE
feat: add support for "down" media queries in "MediaQuery.is()" #11122

### LINKED ISSUE #11122: incorrect Foundation.MediaQuery results
Not sure why but the results seem incorrect.

https://codepen.io/DanielRuf/pen/JLBjBy

### PR DESCRIPTION
**Pull request by ncoden**
## Description

This pull request add support for the "down" keyword in `MediaQuery.is()` thanks to the new method `MediaQuery.upTo()`. It also add `MediaQuery.only()` and unit tests for `MediaQuery.atLeast()`, `MediaQuery.only()` and `MediaQuery.upTo()`.

- Closes https://github.com/zurb/foundation-sites/issues/11122

### Unit tests
The implemented process to test media queries is the following
* **Start a local server on port**
  Local ressources with `file://` cannot be used in an iframe in Chrome. We serve the Foundation repositoryon port `3042`.
* **Dynamically create an iframe.**
  Create an iframe Element, then inject into it the pre-compiled mediaQuery CSS configuration and jQuery/Foundation javascript from the local server.
* **Run tests.**
  Initialize the iframe version of MediaQuery, resize the iframe and run tests.

### Changes
* Validate `MediaQuery.is()` input and returns an error for invalid modifiers
* Add `MediaQuery.upTo()` and `MediaQuery.only()`
* Add support of `down` breakpoint modifier to "MediaQuery.is()`
* Add unit tests for `MediaQuery.atLeat()`, `MediaQuery.only()` and `MediaQuery.upTo()`
* Fix an issue when MediaQuery is initialized twice.
* Improve and update MediaQuery documentation

## Types of changes

- [ ] Documentation
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
      functionality to change)

## Checklist (all required):

- [x] I have read and follow the [CONTRIBUTING](CONTRIBUTING.md) document.
- [x] There are no other pull request similar to this one.
- [x] The pull request title is descriptive.
- [x] The template is fully and correctly filled.
- [x] The pull request targets the right branch (`develop` or `develop-v...`).
- [x] My commits are correctly titled and contain all relevant information.
- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly to my changes (if relevant).
- [x] I have added tests to cover my changes (if relevant).
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by ncoden** [comment]
@DanielRuf @SassNinja for review 👋

File: docs/pages/media-queries.md [hunk 1]
```diff
@@ -197,22 +197,24 @@ Get the name of the current breakpoint with `MediaQuery.current`.
...
-}
+// ↑ True for "medium" or wider (by default)
+Foundation.MediaQuery.is('medium up');
+Foundation.MediaQuery.atLeast('medium');
+
+// → True for "medium" only
+Foundation.MediaQuery.is('medium only');
+Foundation.MediaQuery.only('medium');
+
+// ↓ True for "medium" or smaller
```
Comment (DanielRuf) [inline docs/pages/media-queries.md]: small and wide are two different things. It is either smaller and bigger or narrower and wider.
But not smaller and wider.

File: js/foundation.util.mediaQuery.js [hunk 2]
```diff
@@ -121,20 +123,64 @@ var MediaQuery = {
     return false;
   },
 
+  /**
+   * Checks if the screen is within the given breakpoint.
+   * If smaller than the breakpoint of wider than its upper limit, returns false.
```
Comment (DanielRuf) [inline js/foundation.util.mediaQuery.js]: See my other note about wider / bigger.

Comment (DanielRuf) [reply js/foundation.util.mediaQuery.js]: , => it

File: js/foundation.util.mediaQuery.js [hunk 3]
```diff
@@ -121,20 +123,64 @@ var MediaQuery = {
...
+   */
+  only(size) {
+    return size === this._getCurrentSize();
+  },
+
+  /**
+   * Checks if the screen is within a breakpoint or smaller.
+   * @function
+   * @param {String} size - Name of the breakpoint to check.
+   * @returns {Boolean} `true` if the breakpoint matches, `false` if it's wider.
```
Comment (DanielRuf) [inline js/foundation.util.mediaQuery.js]: wider / bigger

File: js/foundation.util.mediaQuery.js [hunk 4]
```diff
@@ -121,20 +123,64 @@ var MediaQuery = {
...
+   */
+  upTo(size) {
+    const nextSize = this.next(size);
+
+    // If the next breakpoint do not match, the screen is smaller than
```
Comment (DanielRuf) [inline js/foundation.util.mediaQuery.js]: do => does

File: js/foundation.util.mediaQuery.js [hunk 5]
```diff
@@ -121,20 +123,64 @@ var MediaQuery = {
...
+    // the upper limit of this breakpoint.
```
Comment (DanielRuf) [inline js/foundation.util.mediaQuery.js]: the => The

File: js/foundation.util.mediaQuery.js [hunk 6]
```diff
@@ -121,20 +123,64 @@ var MediaQuery = {
...
+    // Only the breakpont
+    if (bpModifier === 'only') {
+      return this.only(bpSize);
     }
-    return false;
+    // Up to the breakpoint (included)
+    if (bpModifier === 'down') {
+      return this.atLeast(bpSize);
+    }
+    // At leat the breakpoint  (included)
```
Comment (DanielRuf) [inline js/foundation.util.mediaQuery.js]: point  (included) = point (included)   (one space less)

File: js/foundation.util.mediaQuery.js [hunk 7]
```diff
@@ -121,20 +123,64 @@ var MediaQuery = {
...
+    if (!bpModifier || bpModifier === 'up') {
+      return this.upTo(bpSize);
+    }
+
+    throw new Error(`
+      invalid breakpoint passed to MediaQuery.is().
+      Expected a breakpoint breakpoint name formatted like "<size> <modifier>", got "${size}".
```
Comment (DanielRuf) [inline js/foundation.util.mediaQuery.js]: breakpoint breakpoint => breakpoint (duplication)

File: js/foundation.util.mediaQuery.js [hunk 8]
```diff
@@ -154,6 +200,43 @@ var MediaQuery = {
...
+   * @function
+   * @param {String} size - Name of the breakpoint.
+   * @returns {String|null} - The name of the following breakpoint, or `null` if the passed breakpoint was the last one.
+   */
+  next(size) {
+    const queryIndex = this.queries.findIndex((q) => this._getQueryName(q) === size);
+    if (queryIndex === -1) {
+      throw new Error(`
+        unknown breakpoint "${size}" passed to MediaQuery.next().
+        Ensure it is present in your Sass settings "$breakpoints".
```
Comment (DanielRuf) [inline js/foundation.util.mediaQuery.js]: settings "$breakpoints" => "$breakpoints" setting

File: js/foundation.util.mediaQuery.js [hunk 9]
```diff
@@ -154,6 +200,43 @@ var MediaQuery = {
...
+   * @param {String|Object} value - Breakpoint name or query object.
+   * @returns {String} Name of the breakpoint.
+   */
+  _getQueryName(value) {
+    if (typeof value === 'string')
+      return value;
+    if (typeof value === 'object')
+      return value.name;
+    throw new TypeError(`
+      invalid value passed to MediaQuery._getQueryName().
```
Comment (DanielRuf) [inline js/foundation.util.mediaQuery.js]: invalid => Invalid

File: test/javascript/lib/utils.js [hunk 10]
```diff
@@ -40,4 +40,29 @@
     }, _opts.interval);
   };
 
+
+  /**
+  * Inject a script with the attributes of the `opts` object into the
+  * given `context` HTML element. The created script is garanted to be executed
```
Comment (DanielRuf) [inline test/javascript/lib/utils.js]: garanted => guaranteed

File: test/javascript/util/mediaQuery.js [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+    it('returns "true" when within the given breakpoint', function () {
+      $iframe.attr('width', 832); // within the "medium" breakpoint
+      (_window.innerWidth); // force the browser to handle the new width synchronously
+
+      plugin.only('small').should.be.false;
+      plugin.only('medium').should.be.true;
+      plugin.only('large').should.be.false;
+    });
+
+    it('returns "false" when wider than given breakpoint', function () {
```
Comment (DanielRuf) [inline test/javascript/util/mediaQuery.js]: than => than the

**Review by DanielRuf — CHANGES_REQUESTED** [review]

File: docs/pages/media-queries.md (hunk 1 above)
Comment (ncoden) [reply docs/pages/media-queries.md]: What about `smaller` and `larger`? This is how the documentation already describe breakpoint.

File: js/foundation.util.mediaQuery.js (hunk 5 above)
Comment (ncoden) [reply js/foundation.util.mediaQuery.js]: No.
> the screen is smaller than (...) the upper limit of this breakpoint.

I break comments at 80 when possible.

**Comment by ncoden** [comment]
Fixed, and resolved conflicts with `develop`

**Comment by andycochran** [comment]
@DanielRuf Do you approve the additional commits @ncoden made after your review? 

Might you be able to resolve the conflicts? Looks like most of them could be resolved by trashing the lock files and reinstalling dependencies.

**Comment by DanielRuf** [comment]
I'll try to find some time to answer in the next few days.

**Review by DanielRuf — APPROVED** [review]
LGTM

**Comment by joeworkman** [comment]
This pull request has been mentioned on **Foundation Open Source Community**. There might be relevant details there:

https://foundation.discourse.group/t/foundation-for-sites-v6-6-0-is-here-farout/30/1

**Pull request merged**